jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt

```diff
@@ -40,51 +39,84 @@ interface FileCrawler {
+     * (4) writable file
      */
-    fun listRelevantFilesInEditors(psiFile: PsiFile): List<VirtualFile>
+    fun listCrossFileCandidate(psiFile: PsiFile): List<VirtualFile>
+
+    /**
+     * Determine if the file given is test file or not based on its path and file name
+     */
+    fun isTestFile(virtualFile: VirtualFile, project: Project): Boolean
 }
 
 class NoOpFileCrawler : FileCrawler {
     override suspend fun listFilesImported(psiFile: PsiFile): List<VirtualFile> = emptyList()
 
     override fun listFilesUnderProjectRoot(project: Project): List<VirtualFile> = emptyList()
-    override fun findFocalFileForTest(psiFile: PsiFile): VirtualFile? = null
+    override fun listUtgCandidate(psiFile: PsiFile): VirtualFile? = null
 
     override fun listFilesWithinSamePackage(psiFile: PsiFile): List<VirtualFile> = emptyList()
 
-    override fun listRelevantFilesInEditors(psiFile: PsiFile): List<VirtualFile> = emptyList()
+    override fun listCrossFileCandidate(psiFile: PsiFile): List<VirtualFile> = emptyList()
+
+    override fun isTestFile(virtualFile: VirtualFile, project: Project): Boolean = false
 }
 
 abstract class CodeWhispererFileCrawler : FileCrawler {
     abstract val fileExtension: String
-    abstract val testFilenamePattern: Regex
     abstract val dialects: Set<String>
+    abstract val testFileNamingPatterns: List<Regex>
+
+    override fun isTestFile(virtualFile: VirtualFile, project: Project): Boolean {
+        val filePath = virtualFile.path
+
+        // if file path itself explicitly explains the file is under test sources
+        if (TestSourcesFilter.isTestSources(virtualFile, project) ||
+            filePath.contains("""test/""") ||
+            filePath.contains("""tst/""") ||
+            filePath.contains("""tests/""")
+        ) {
+            return true
+        }
+
+        // no explicit clue from the file path, use regexes based on naming conventions
+        return testFileNamingPatterns.any { it.matches(virtualFile.name) }
+    }
 
     override fun listFilesUnderProjectRoot(project: Project): List<VirtualFile> = project.guessProjectDir()?.let { rootDir ->
         VfsUtil.collectChildrenRecursively(rootDir).filter {
+            // TODO: need to handle cases js vs. jsx, ts vs. tsx when we enable js/ts utg since we likely have different file extensions
             it.path.endsWith(fileExtension)
         }
     }.orEmpty()
 
-    override fun listRelevantFilesInEditors(psiFile: PsiFile): List<VirtualFile> {
-        val targetFile = psiFile.virtualFile
-        val language = psiFile.programmingLanguage()
+    override fun listFilesWithinSamePackage(psiFile: PsiFile): List<VirtualFile> = runReadAction {
+        psiFile.containingDirectory?.files?.mapNotNull {
+            // exclude target file
+            if (it != psiFile) {
+                it.virtualFile
+            } else {
+                null
+            }
+        }.orEmpty()
+    }
```

Comment:
moved from https://github.com/aws/aws-toolkit-jetbrains/pull/3774/files#diff-220167937b0378fbb221953b4f601e646f0519d2013469c04cd1cac97f25a7bdL70, but this method is not being used for now (only used in tests).